File: modules/rates-widget/android/src/main/java/expo/modules/rateswidget/RatesWidgetModule.kt
package expo.modules.rateswidget

import android.content.Context
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class RatesWidgetModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("RatesWidget")

    AsyncFunction("refreshWidgets") {
      RatesWidgetWorker.enqueueImmediate(context)
    }

    AsyncFunction("getCachedRates") {
      RatesWidgetRepository.getCachedRates(context)?.toMap()
    }
  }
}

File: modules/rates-widget/src/RatesWidgetModule.ts
import { NativeModule, requireNativeModule } from 'expo';
import type { WidgetRates } from './RatesWidget.types';

declare class RatesWidgetModule extends NativeModule<{}> {
  refreshWidgets(): Promise<void>;
  getCachedRates(): Promise<WidgetRates | null>;
}

export default requireNativeModule<RatesWidgetModule>('RatesWidget');
